kite-data/kite-data-core/src/test/java/org/kitesdk/data/spi/filesystem/TestAvroWriter.java

```diff
@@ -16,17 +16,80 @@
 
 package org.kitesdk.data.spi.filesystem;
 
+import com.google.common.collect.Lists;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericData.Record;
+import org.apache.hadoop.fs.FileStatus;
 import org.apache.hadoop.fs.Path;
+import org.junit.Assert;
+import org.junit.Test;
 import org.kitesdk.data.DatasetDescriptor;
+import org.kitesdk.data.DatasetReader;
 import org.kitesdk.data.DatasetWriter;
+import org.kitesdk.data.Flushable;
+import org.kitesdk.data.Syncable;
+import org.kitesdk.data.spi.ReaderWriterState;
 
-public class TestAvroWriter extends TestFileSystemWriters<Object> {
+public class TestAvroWriter extends TestFileSystemWriters {
   @Override
-  public DatasetWriter<Object> newWriter(Path directory) {
-    return new FileSystemWriter<Object>(fs, directory,
+  public FileSystemWriter<Record> newWriter(Path directory, Schema schema) {
+    return FileSystemWriter.newWriter(fs, directory,
         new DatasetDescriptor.Builder()
-            .schemaLiteral("\"string\"")
+            .schema(schema)
             .format("avro")
             .build());
   }
+
+  @Override
+  public DatasetReader<Record> newReader(Path path, Schema schema) {
+    return new FileSystemDatasetReader<Record>(fs, path, schema, Record.class);
+  }
+
+  @Test
+  public void testIsFlushable() {
+    Assert.assertTrue(fsWriter instanceof FileSystemWriter.IncrementalWriter);
+  }
+
+  @Test
+  public void testIsSyncable() {
+    Assert.assertTrue(fsWriter instanceof FileSystemWriter.IncrementalWriter);
+  }
+
+  @Test
+  public void testCommitFlushedRecords() throws IOException {
+    init(fsWriter);
+
+    List<Record> written = Lists.newArrayList();
+    long i;
+    for (i = 0; i < 10000; i += 1) {
+      Record record = record(i, "test-" + i);
+      fsWriter.write(record);
+      written.add(record);
+    }
+
+    ((FileSystemWriter.IncrementalWriter) fsWriter).flush();
+
+    for (i = 10000; i < 11000; i += 1) {
+      fsWriter.write(record(i, "test-" + i));
+    }
+
+    // put the writer into an error state, simulating either:
+    // 1. A failed record with an IOException or unknown RuntimeException
+    // 2. A failed flush or sync for IncrementableWriters
+    fsWriter.state = ReaderWriterState.ERROR;
+
+    fsWriter.close();
+
+    FileStatus[] stats = fs.listStatus(testDirectory, PathFilters.notHidden());
+    Assert.assertEquals("Should contain a visible data file", 1, stats.length);
+
+    // TODO: This currently fails because the file needs to be recovered
```

Comment:
Yes. This is failing because we're using the local filesystem, and hflush() isn't implemented for it (https://issues.apache.org/jira/browse/HADOOP-7844). I tried switching TestFileSystemWriters to extend MiniDFSTest and the commented out test passes.